support reading rustc arguments from a file

TimNN opened the pull request:
Add an unstable (for now) flag to rustc (--include-arguments-from FILE) which causes rustc to read additional arguments (\0 separated) from the specified FILE or stdin if FILE is '-'.

I consider this pull request still a work-in-progress, because it's currently lacking any kind of test, but I wanted to publish it now to get some feedback, mainly on whether the current approach is the correct one and if this kind of change would require an RFC.

The pull request was inspired by rust-lang/cargo#1241.

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented:
I'll take a look at this more closely tomorrow

retep998 commented:
Yay!

TimNN commented:
One thing that I just noticed and which may require special consideration is that `--` processing is currently "scoped" to the current include (a `--` in an include will only remain in effect for that include).

nagisa commented on src/librustc/session/config.rs:
```diff
@@ -841,6 +844,10 @@ pub fn rustc_optgroups() -> Vec<RustcOptGroup> {
+        opt::include_u("", "include-arguments-from",
+                       "Read additional arguments from the given FILE or the \
+                        standard input if FILE is `-`. Arguments must be \
+                        separated by \\0.", "FILE", b'\0'),
```
There’s ASCII control characters dedicated for separating things (namely – `\u{1E}` for separating different records (flag-value pairs) from each other). NULL byte is not one of those.

TimNN replied:
`\0` was mentioned by @huonw in rust-lang/cargo#1241, which is why I choose to use it, but I can change that easily if required.

huonw replied:
I've personally never seen anything other than `,`, `;`, `:`, whitespace (spaces and/or newlines) or `\0` used to delimit records. That said, if there's a dedicated character for it...

nagisa commented:
I’d want this option to live under `-Z whatever-name`.

alexcrichton commented:
Ok, this looks pretty good to me, and here's my thoughts on an option like this:

* I think an option along these lines is inevitable and we're gonna need to add it to the compiler at some point. The linked Cargo issue is a legitimate use case and I have a feeling this restriction will start to pop up more often.
* There is prior art here in a number of GNU tools that take an argument of the form `@file` where `file` is read for options. The main difference seem to be that GNU uses whitespace as a separator, including quotes/backslashes in the syntax to parse the options with spaces.
* I like having a dedicated flag for this instead of special `@` syntax like GNU. I don't think that anyone's going to be actually using this sort of option manually, I expect the only real use case to be tooling like Cargo or other `rustc` drivers possibly. Along those lines I don't think we need a sugary way of specifying this option.
* I'm ok with a byte terminator between arguments. They're very hard to generate by hand but along the lines of the previous bullet I don't think that's the use case this sort of option is catering to, I expect it to largely just be tooling.
* I'm ok with the nul byte as a terminator, it's easy to remember and rarely shows up in arguments (e.g. it's *very* unlikely to have a nul byte in a path).
* Don't worry too much about dealing with `--`, this library is only really used by our own internal tools and I believe none of them need that kind of functionality.
* Treating `-` as stdin can probably be elided for now, I don't think the compiler handles this for output and along the lines of the points above I don't think many tools will be piping options to the compiler via this mechanism (but will instead just write out files).
* This is a new option to the compiler, and despite being behind `-Z unstable-options` (which isn't really gated at all) it's semi-insta-stable, so I'd want to tread carefully.

cc @rust-lang/tools, this is a new argument to the compiler which will probably make its way into at least `rustdoc` at some point if not other places.

alexcrichton commented:
Ah! And can you be sure to add a number of tests for this? I'd like to see a few smoke tests and at least one for the recursive include case.

nagisa commented:
> Treating - as stdin can probably be elided for now, I don't think the compiler handles this for output and along the lines of the points above I don't think many tools will be piping options to the compiler via this mechanism (but will instead just write out files).

I don’t see how otherwise `cargo` would reliably pass in arguments to rustc. Creating a temporary file for rustc to read these off it would be racy when there’s more than one concurrent invocation of rustc…

Well, you could create a temporary file for each rust instance, but do you really want to create that many temporaries?

alexcrichton commented:
Yeah Cargo would just create a file per rustc invocation, it's already got lots of unique places to put the files and there's already tons of files flying around so I don't expect this to ever be a real problem. I also plan to not actually use this unless Cargo determines the argument list is too long as it hurts the debuggability of `cargo build -v` output by default.

arielb1 commented:
@nagisa

On Linux you can always pass "/proc/self/fd/<fd>" to a pipe. I think there is also a Windows equivalent.

retep998 commented:
@arielb1 `\\.\pipe\mynamedpipe`

vadimcn commented:
Is there a strong reason to prefer `\0` to `\n`?   Yes, this is not intended for humans, but it'd still be nice if a parameter file could be opened in a regular editor, if only to inspect its contents.

TimNN commented:
There *may* be occasions where a `\n` needs to be included in some argument, thus there would need to be some way to escape a `\n` which would make the whole parsing more complicated.

TimNN commented:
Just a quick update / some of my thoughts:

* I briefly considered the `@filename` syntax, but decided that it would be cleaner, more elegant and practical to use a real flag.
* As mentioned in my previous comment I deliberately decided against some kind of whitespace separator because it would have made the parsing more difficult.
* I removed the possibility to read from stdin using `-`, which simplified the code a little bit and removed the need to write tests for that functionality.
* I'm currently writing some tests but it may take a few days before I find the time to finish them. 

Which leads me to a question: What's the policy of using internal `rustc_private` crates for tests? I noticed that the existing `getopts` tests already use the `log` crate (which seems kinda stable) and am currently using `rustc_back::tempdir::TempDir` (which does not seem to be very widely used) in the tests to get a location where I can write the temporary files to.

alexcrichton commented:
Ah yeah feel free to use `rustc_private` however you'd like, ideally these tests would be in the `getopts` crate itself, but it's fine to have them live elsewhere in `run-pass`.

One other thing I just remembered is that this would be a relatively large departure from the current `getopts` crate on crates.io, which is otherwise the same as the in-tree variant with just a slightly different API and some assorted bugs fixed.

@vadimcn does have a good point that reading the files would be easier with `\n`, and I don't think it's necessarily harder to parse (e.g. `split(b'\n')` instead of `split(0)`), so it's worth considering. It's true that I doubt anything actually has a newline in the argument, and we could always add some form of escaping perhaps.

retep998 commented:
For the record, pretty much all the flags that will be passed are going to be paths or will be used in paths, and on Windows "Characters whose integer representations are in the range from 1 through 31" are not allowed in paths, so newlines and any of those other control characters are all totally fine to use.

TimNN commented:
If you prefer, I'm fine with changing to `\n` as the separator, but think that the inspecting / editor use case is easily available with something like `cat file_with_nul | tr "\0" "\n"`.

Would you consider the escaping part of this PR or as part of a followup PR should it be needed?

Are the in-tree and crates.io versions of crates like getopts usually kept in sync? If so, should I merge the crates.io changes back into the in-tree version so my changes can more easily be merged into the crates.io version as well?

alexcrichton commented:
I think there's certainly an appeal to not having to go through a postprocessing step to get a readable file (e.g. in favor of newlines), but I also don't personally feel particularly strong one way or another. I think we may want to consider escaping here in terms of design, and that'd influence whether it needs to be implemented. For example I don't think that as-is we could add escaping backwards compatibly to a format that's "just newline separated" as whatever the escape character is would need to also be escaped.

As for the in-tree vs crates.io crate, they're not normally intentionally kept in sync (the external one tends to evolve though). I only bring this up in the sense that one day we'd like to use external crates in our build here instead of having vendored versions. I think it's fine to basically decide what to do here and then perhaps move it to the external crate as well.

TimNN commented:
So, let's consider the implications of choosing to use newlines and add escaping:
* The usual escape character is `\` which is what we would use I imagine
* To keep the "format" simple we could define `\` to mean that any special meaning of the following character will be ignored. `\n` would thus be a literal `n`, `\\`a literal `\` and `\<LINEBREAK>` a literal line break and not an argument separator.
* The sequence `\<EOF>` would be considered an error.
* The parsing would be a little bit more complicated (read the input `char` by `char`).
* Cargo (or any other program using this feature) would have to escape any `\` in the arguments (such as Windows paths).
* With the editor use case, carefully consider what to do about an empty last argument (resulting from a file ending with a new line [which I, for example, configured my editors to do automatically): without a special case this would lead to an empty argument which could result in an error but also be intended.
* Be careful with the language in any documentation: this would be `LF` separated, not any other line ending.

Some more thoughts about actually using `LF` as the argument separator and the editor use case:
* When I originally wrote this PR this feature was intended to be mainly / only be used by automated tools which require passing lots of arguments to `rustc`. 
* I imaging that any tool would clean up any temporary files soon after the `rustc` invocation, making it difficult to get hold of the file with the arguments.
* I also imaging that any tool would provide the ability to log the `rustc` invocation in other ways.
* At least the usual linux editors (vim, emacs, nano) are perfectly capable of rendering `NUL`, other editors vary (Sublime displays the file as hex, TextEdit and Atom don't display `NUL` but preserve it during edit / save, Notepad++ will render `NUL` if I remember correctly).

Considering all of that I think we should decide between the "Machine Readable" use case, using `NUL` as the separator or the "Human Readable / Editable" use case and use a [gcc like](https://gcc.gnu.org/onlinedocs/gcc-4.9.0/gcc/Overall-Options.html) (search for `@`), whitespace separated syntax.

I would define the whitespace separated syntax like this:
* All whitespace characters are considered argument separators
* `\` escapes the following character as above.
* Any consecutive whitespace is collapsed, whitespace at the beginning and end of file is ignored.
* If the first character of an argument is a `"` or `'`, all characters up to the next matching `"` or `'` are considered part of the argument.

Using the whitespace separated syntax would of course make it more difficult for any tools to properly escape the arguments it want's to pass to `rustc`.

Personally, I would prefer going the machine readable route but would have not problem implementing the whitespace based approach.

retep998 commented:
Don't forget that with paths on Windows using `\`, that means you'd have to do a _lot_ of escaping via `\\`.

arielb1 commented:
\0-separated records is a standard unix format, and easy to translate to/from newline-separated escaped.

TimNN commented:
Considering all of the discussion I decided to stick with `\0` as the argument separator for the moment. If something else is desired, please let me know.

Apart from that I added some tests (in `run-pass` due to the `rustc_back::tempdir` dependency), which test the support for arbitrary characters in the arguments and some corner cases. If you think of anything else which needs to be tested, please let me know.

Pending a successful travis run, I would consider this PR complete and ready to be merged.

TimNN commented:
`check-stage1-rpass TESTNAME=getopts` ran without problems locally, so lets try and figure out why stage2 failed `>.<`.

TimNN commented:
Moving the test to `run-pass-fulldeps` fixed the problem locally and I imagine that travis will be happy as well :smile:

brson commented:
Not sure how I feel about this yet.

That cargo wants such ginormous command line is a smell. There may be alternative approaches that support cargo's use cases. For example, most of these command line arguments are taken up by mapping crates to libraries, and *this specifically* could be driven by a file without imposing this generic command line parsing facility on not just rustc but all future rustc-compatible compilers. For another example, if cargo just used rustc as a library this would not be an issue.

The implementation here is complex compared to the value added - this is basically a hack to support cargo, but the semantics here of *mixing* in arguments from another source seems to be imposing complexity when cargo doesn't need that - it can put all arguments in one file.

That this is directly modifying the getopts library also bothers me - this appears an unconventional command line parsing strategy; getopts is old and deprecated everywhere but in rustc and somebody will want to move away  from it eventually; the weirder it is the harder that will be.

alexcrichton commented:
Closing due to inactivity for now, I sympathize with @brson's concerns and I also worry that if the rustc command line is too big it's almost guaranteed to mean the linker command line is *also* too big (which isn't necessarily handled here). Along those lines, this sort of situation may need to be patched up in more than a few places.

It's also the case that the corresponding Cargo bug isn't necessarily reflective of idiomatic usage today, so while it'd be great to fix it's probably not critical. Regardless though, @TimNN if you'd like to move forward feel free to reach out to me/@brson in `#rust-tools` on IRC and we may be able to help push this forward!

TimNN commented:
I'm fine with closing this, I was planning on doing so anyway in the next few days due to lack of consensus / enthusiasm.
